### ISSUE TITLE
Updating to from .NET 8 SDK to .NET 9 SDK Preview 4 causes `dotnet test` to hang forever

### ISSUE THREAD

**Issue by Youssef1313** [microsoft/vstest#5091]
## Description

So far, I don't have an isolated repro. But it happens for Uno Platform Wasm UI tests that we execute via `dotnet test`. The tests are passing, but `dotnet test` isn't terminating.

Upon investigation, I found:

![image](https://github.com/microsoft/vstest/assets/31348972/e182d955-da31-4c70-9509-008128983508)

![image](https://github.com/microsoft/vstest/assets/31348972/5412aa20-39fd-4d5a-97f0-d67ce729ffc9)

So it looks like somehow, the `VSTestTask2` isn't terminating. It's stuck there forever. Setting `MSBUILDENSURESTDOUTFORTASKPROCESSES` environment variable to `1` does the trick for now.

## Steps to reproduce

We haven't yet got a minimal repro.

## Diagnostic logs

[logs.zip](https://github.com/user-attachments/files/15787756/logs.zip)

**Comment by nohwnd** [comment]
vstesttask2 task starts an exe, and waits for it to exit. It will sit there as long as the exe will be running. When you look in test explorer do you see vstest.console / dotnet running under this process? Do you see also testhost running under vstest.console process?

**Comment by Youssef1313** [comment]
> vstesttask2 task starts an exe

Is it the `testhost.exe`? That one terminates correctly

**Comment by Youssef1313** [comment]
I have a dump of the `dotnet` process, in case that can help, I'll send it to you.

**Comment by nohwnd** [comment]
I've got your dump, it looks like the tool task is simply waiting for a child process to exit. The child process is vstest.console.

In the logs of vstest.console I can see it exited, but I also see that the Process ID is different from what is in the dump file, so this is probably from 2 different runs, not a big problem, but could you double check that vstest.console is stopped under the task? if you put long wait in your test, you should see one under some MSBuild node, and then it should exit. 

You could also try using -nodereuse:false, that will disable using "cached" MSbuild nodes, and will start a new one for this run, if it still stays stuck it makes it much easier to see what process is stuck, because they all run under the terminal process.

**Comment by nohwnd** [comment]
That is indeed weird, and in that case it would be a MSBuild bug (not that I am trying to ditch responsibility, but we are fully relying on ToolTask to do this). Let me know what you found and I will talk with msbuild team if there is problem in tooltask.

**Comment by Youssef1313** [comment]
Great. I'll double check my analysis and try to get more info and get back to you

**Comment by tmds** [comment]
We ran into this issue in https://github.com/dotnet/sdk/pull/41198 CI.

The CI test step would time out after 30 min on each attempt.
After adding `MSBUILDENSURESTDOUTFORTASKPROCESSES=1` the test step finishes in less than 4 min.

cc @ViktorHofer @rainersigwald @Forgind

**Comment by Forgind** [comment]
On a hunch, can someone try setting MSBUILDNODEWINDOW to 1 in addition to MSBUILDENSURESTDOUTFORTASKPROCESSES=1 and see if the problem comes back?

**Comment by Forgind** [comment]
Actually, I guess I can do that. I'll try to get that started later today.

**Comment by nohwnd** [comment]
This is now on list of work for msbuild team, and me to fix. We still don't know where it is happening though. So if you have any additional info, or repro it would be very welcome. Especially double checking if vstest.console is or is  not running while the hang is observed. And diagnostic logs of dotnet test.

**Comment by MichalPavlik** [comment]
There is a thing with `WaitForExit()` method when parent process reads stdout asynchronously. If there is a grandchild process started by child process, `WaitForExit()` of the parent process waits for exit of the grandchild. It blocks even when the child process exits.

I'm not saying it's the root cause in this situation, but it's possible. Our `ToolTask` uses `WaitForExit()`, so I can try to avoid this situation on our side.

**Comment by Youssef1313** [comment]
Not sure if that would be related to https://github.com/dotnet/runtime/issues/103384

**Comment by MichalPavlik** [comment]
What I described relates to issue you mentioned. @Youssef1313, could you please try to find if there is a process that was started by the `testhost` and terminate it? If it unblocks the MSBuild, then the problem is in our codebase and should be fixed. The workaround is to use different overload of `WaitForExit` method.

**Comment by Youssef1313** [comment]
I may not be able to re-test this soon-ish, but IIRC, when WaitForExit was stuck, I wasn't able to find a matching process id that's open. So it felt like that process already terminated but WaitForExit was still blocking and didn't return.

**Comment by MichalPavlik** [comment]
Yes, if `testhost` started another process with redirected output, then our `WaitForExit` will wait for the grandchild process to exit. So far I don't have another idea what is happening.

**Comment by MichalPavlik** [comment]
I implemented workaround but I had to revert my changes, because it caused problem with exit code :( Still, it would be great to know if there is hanging grandchild process. In that case, testing team could make change here and don't exit before their child process terminates.

**Comment by Sebazzz** [comment]
This indeed happens when the test itself creates child processes (for instance: MSSQL localdb, or webdrivers/browsers) and fails to terminate these child processes. `dotnet test` will then indeed hang. 

Still happens with SDK 9.0.100.

**Comment by nohwnd** [comment]
Looks to be connected to this change: https://github.com/microsoft/vstest/pull/4998 where we started capturing output from child processes.

Adding `VSTEST_DISABLE_STANDARD_OUTPUT_CAPTURING=1` env variable to my build unblocked the run.

https://github.com/nohwnd/Cuemon/blob/3126ca48266b9ad493524a06ea9b77c79e907a30/.github/workflows/pipelines.yml#L157-L158

**Comment by nohwnd** [comment]
Actually no matter what I do I cannot repro the hang when running on 9.0.100, so I am not sure if the option above did anything.

**Comment by nohwnd** [comment]
OP confirmed this is working for them now.

 @Sebazzz do you have a repro of your problem please? If it still does repro, please start a new issue.
